packages/ic-http-gateway/src/protocol/validate.rs

```diff
@@ -0,0 +1,61 @@
+use crate::IC_CERTIFICATE_HEADER_NAME;
+use candid::Principal;
+use ic_agent::Agent;
+use ic_http_certification::{HttpRequest, HttpResponse};
+use ic_response_verification::{
+    types::VerificationInfo, verify_request_response_pair, MIN_VERIFICATION_VERSION,
+};
+use std::{
+    borrow::Cow,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+const MAX_CERT_TIME_OFFSET_NS: u128 = 300_000_000_000;
+
+pub fn validate(
+    agent: &Agent,
+    canister_id: &Principal,
+    request: HttpRequest,
+    response: HttpResponse,
+    allow_skip_verification: bool,
+) -> Result<Option<VerificationInfo>, Cow<'static, str>> {
+    match (allow_skip_verification, has_ic_certificate(&response)) {
+        // TODO: Remove this (FOLLOW-483)
+        // Canisters don't have to provide certified variables
+        // This should change in the future, grandfathering in current implementations
+        (true, false) => Ok(None),
+        (_, _) => {
+            let ic_public_key = agent.read_root_key();
+            let verification_info = verify_request_response_pair(
+                request,
+                response,
+                canister_id.as_slice(),
+                get_current_time_in_ns(),
+                MAX_CERT_TIME_OFFSET_NS,
+                ic_public_key.as_slice(),
+                MIN_VERIFICATION_VERSION,
+            )
+            .map_err(|_| "Body does not pass verification")?;
```

Comment:
maybe log here the actual error that response verification has thrown

Reply:
I agree we should do something here. Let's discuss more and handle in a separate PR though.